When serializing treespec context, support enum as well

henrylhtsang opened the pull request:
Following https://github.com/pytorch/pytorch/pull/102716, per @angelayi's suggestion.

Note that in general enum as an input is not supported.

repro:
```
class TestEnum(enum.Enum):
    A = auto()
    B = auto()

    @staticmethod
    def from_string(s):
        return TestEnum[s.upper()]


class M(torch.nn.Module):
    def forward(self, x, en):
        return x.clone()

input1 = (
    torch.rand(10, device="cuda"),
    {TestEnum.A: torch.rand(10, device="cuda")},
)
inputs = [input1]
model = M().cuda()

_ = model(*input1)

ep = torch.export.export(model, input1, strict=False)
path = torch._inductor.aot_compile(ep.module(), input1)
```

Differential Revision: D66269157

angelayi reviewed (commented):
can you add a test case into test/test_pytree.py?

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -1369,7 +1377,7 @@ def _treespec_to_json(treespec: TreeSpec) -> _TreeSpecSchema:
     if serialize_node_def.to_dumpable_context is None:
         try:
-            serialized_context = json.dumps(treespec.context)
+            serialized_context = json.dumps(treespec.context, cls=EnumEncoder)
```
How does one end up with an enum in the treespec's context?

henrylhtsang replied:
@zou3519 I think the forward function takes in a dict, which looks like `{some enum: torch.rand(3)}`

zou3519 replied:
Do exported programs accept raw enums as inputs? If the answer is no, then they should also not accept dicts that have enums as keys as inputs.

henrylhtsang replied:
@zou3519 export works. Problem seems to appear when calling aot_compile

angelayi replied:
Export just treats enums as constants, so this seems like something we could support?

zou3519 replied:
It seems like it was a bad idea to give pytrees their own serialization mechanism; a better design is looking like that export should have its own pytree serialization mechanism that can be tied to how export works. For example, the enum thing: the serialization here makes sense as long as obj.value is serializable, if it's not serializable then we've got a problem and there's no way for a generic user to specify it.

I don't know what we want to do with this information. I'm happy to approve to unblock this change

angelayi replied:
> a better design is looking like that export should have its own pytree serialization mechanism that can be tied to how export works

Yeah this makes sense. Made an issue to track: https://github.com/pytorch/pytorch/issues/141996

henrylhtsang committed:
When serializing treespec context, support enum as well (#141525)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/141525

Test Plan: test

Differential Revision: D66269157

henrylhtsang commented:
@pytorchbot merge -i
